src/uct/ib/base/ib_device.c

```diff
@@ -408,6 +428,109 @@ ucs_status_t uct_ib_device_port_check(uct_ib_device_t *dev, uint8_t port_num,
     return UCS_OK;
 }
 
+static int uct_device_get_addr_family(union ibv_gid *gid, int gid_index)
+{
+    const struct in6_addr *raw = (struct in6_addr *)gid->raw;
+    char p[128];
+
+    ucs_debug("testing addr_family on gid index %d: %s",
+              gid_index, inet_ntop(AF_INET6, gid, p, sizeof(p)));
+
+    if (((raw->s6_addr32[0] | raw->s6_addr32[1]) |
+         (raw->s6_addr32[2] ^ htonl(0x0000ffff))) == 0UL ||
+        /* IPv4 encoded multicast addresses */
+        (raw->s6_addr32[0] == htonl(0xff0e0000) &&
+         ((raw->s6_addr32[1] | (raw->s6_addr32[2] ^ htonl(0x0000ffff))) == 0UL))) {
```

Comment:
This calculation can be separate function to avoid code duplication and better readability